versioned_hdf5/slicetools.pyx

```diff
@@ -0,0 +1,178 @@
+        slices: list = []
+
+        rank: cython.int = H5Sget_simple_extent_ndims(space_id)
+        start_array: cython.pointer(hsize_t) = <hsize_t *> malloc(sizeof(hsize_t) * rank)
+        stride_array: cython.pointer(hsize_t) = <hsize_t *> malloc(sizeof(hsize_t) * rank)
+        count_array: cython.pointer(hsize_t) = <hsize_t *> malloc(sizeof(hsize_t) * rank)
+        block_array: cython.pointer(hsize_t) = <hsize_t *> malloc(sizeof(hsize_t) * rank)
+        try:
+            H5Sget_regular_hyperslab(space_id, start_array, stride_array,
+                                     count_array, block_array)
+            i: cython.int
+            start: hsize_t
+            end: hsize_t
+            stride: hsize_t
+            count: hsize_t
+            block: hsize_t
+            for i in range(rank):
+                start = start_array[i]
+                stride = stride_array[i]
+                count = count_array[i]
+                block = block_array[i]
+                if not (block == 1 or count == 1):
+                    raise NotImplementedError("Nontrivial blocks are not yet supported")
+                end = start + (stride * (count - 1) + 1) * block
+                stride = stride if block == 1 else 1
+                slices.append(Slice(start, end, stride))
+        finally:
+            free(start_array)
+            free(stride_array)
+            free(count_array)
+            free(block_array)
+        return Tuple(*slices)
+    elif sel_type == H5S_sel_type.H5S_SEL_NONE:
+        return Tuple(
+            Slice(0, 0),
+        )
+    else:
+        raise NotImplementedError("Point selections are not yet supported")
+
+cpdef build_data_dict(dcpl, shape: tuple, chunks: tuple, raw_data_name: str):
+    data_dict: dict = {}
+
+    is_virtual: bool = dcpl.get_layout() == h5d.VIRTUAL
+
+    if not is_virtual:
+        # A dataset created with only a fillvalue will be nonvirtual,
+        # since create_virtual_dataset makes a nonvirtual dataset when
+        # there are no virtual sources.
+        data_dict = {}
+    # Same as dataset.get_virtual_sources
+    elif 0 in shape:
+        # Work around https://github.com/h5py/h5py/issues/1660
+        empty_idx = Tuple().expand(shape)
+        data_dict = {empty_idx: Slice()}
+    else:
+        data_dict = {}
+
+        with phil:
+            dcpl_id: hid_t = dcpl.id
+
+            virtual_count: size_t = dcpl.get_virtual_count()
+            j: size_t
+
+            raw_data_name_bytes: bytes = raw_data_name.encode('utf8')
+            # this a reference to the internal buffer of raw_data_name, do not free!
+            raw_data_str: cython.p_char = raw_data_name_bytes
+
+            filename_buf_len: ssize_t = 2
+            filename_buf: cython.p_char = <char *>malloc(filename_buf_len)
+
+            try:
+                dataset_buf_len: ssize_t = strlen(raw_data_str) + 1
+                dataset_buf: cython.p_char = <char *>malloc(dataset_buf_len)
```

Comment:
Same question about malloc as above :point_up: - do we need to raise an exception if this fails?